### ISSUE TITLE
LV2 support

### ISSUE THREAD

**Issue by ksnd** [LMMS/lmms#562]

**Comment by tresf** [comment]
Bumping to 1.3 (really 2.0) since it will get lost in 1.2. :pizza:

**Comment by tresf** [comment]
@badosu wrote Mar 10:

> @tresf Well, unfortunately this is totally alien for me, so it will take a while :-). I made some progress building a generic QT host for lv2 plugins, adapting Jalv, see:<br>

> http://bado.so/2015-03-06-031713_1919x1045_scrot.png
![image](https://cloud.githubusercontent.com/assets/6345473/6569808/ea1bcbec-c6c6-11e4-9b93-d9fb5bb3d11f.png)<br>


> This is [amSynth](https://code.google.com/p/amsynth/) on the jalv.qt host that I am helping develop, this will be useful as the analogue for the "controls" menu we use on VSTs to access parameters hidden by the UI or that need automation.

:)

**Comment by Gps2010** [comment]
I like the third this request, because I just found out, I could use calf studio plugin in lmms if we had an lv2 host.

**Comment by Reaper10** [comment]
Is the Linux sampler going added to LMMS?

**Comment by Reaper10** [comment]
Will you be replacing the LMMS versions of the CALF plugings  better GUI of the LV2 versions of the  CALF plugings where you can? After all you don't need two versions of the CALF plugings.

**Comment by tresf** [comment]
Yes, although if backwards compatibility is an issue, we may decide to keep both for a while (all speculative, LV2 support is a ways out)

**Comment by Reaper10** [comment]
I found LV2 versions of the MDA Piano VSTIs ----->   http://git.elephly.net/gitweb.cgi?p=software/lv2-mdametapiano.git

**Comment by michaelgregorius** [comment]
@Reaper10 In my opinion it does not make sense to build everything and the kitchen sink right into LMMS. There's a reason that plugins are called "plugins". :wink:

**Comment by Reaper10** [comment]
There's a reason that plugins are called "plugins". This is true but LMMS could a acoustic piano plugin build in. Besides that Vst are extremely loading.

**Comment by Umcaruje** [comment]
>There's a reason that plugins are called "plugins". This is true but LMMS could a acoustic piano plugin build in. Besides that Vst are extremely loading.

Then we should improve the plugin support and not make more native plugins.

**Comment by michaelgregorius** [comment]
@Reaper10 If LV2 support is implemented (which is what this issue is about) then you should be able to install the piano plugin you have linked and simply use it. Pulling all kinds of plugins and effects into LMMS' code base puts a large burden on the developers to keep them all up to date. Let's not turn LMMS into a Linux audio distribution. :wink:

**Comment by suhr** [comment]
What's the status of this issue?

**Comment by grejppi** [comment]
@suhr personally I'm unwilling to work on this before 1.2 is out of the way

**Comment by musikBear** [comment]
I know that this
https://www.youtube.com/watch?v=ZsgA04tfiwc
is in no way a 'solution', but just want to mention it, because it does work. An option *Export-all-button-settings-to-automation* on this ui:
https://lmms.io/wiki/images/0/08/Synth1_dial_array.png
Would make it easy.
Other 'solutions' of same kind can be found here:
https://www.google.dk/search?q=redneck+solutions+pictures&tbm=isch&tbo=u&source=univ&sa=X&ved=0ahUKEwja08vk4IPLAhXIDQ4KHWmBAjkQsAQIGw&biw=1007&bih=635
:black_joker:

**Comment by simonvanderveldt** [comment]
@grejppi is there no way to implement the LV2 support you worked on without #2616?
Or implement LV2 support on the work in #2616 and keep the existing LADSPA and VST implementations as is for now and then later port over the LADSPA and VST implementations?

Trying to look at ways to keep the changes/improvements small so these important/big items move forward, even if it's just with small steps.

**Comment by grejppi** [comment]
@simonvanderveldt 

> is there no way to implement the LV2 support you worked on without #2616?

Kind of, yes, but there's also absolutely no point in it, because without #2616 one would have to implement all the common functionality one more time for LV2. The reason for #2616 is having a solid abstraction that can be made to support any plugin format.

>  keep the existing LADSPA and VST implementations as is for now and then later port over the LADSPA and VST implementations?

That is a possibility, however, once the new system is in place porting them should be straightforward. (The reason why I started with LADSPA is because it's in fact very simple to implement, yet LMMS currently does that in a very complicated way.)

**Comment by simonvanderveldt** [comment]
@grejppi Thanks for the clarifications, makes sense on both accounts :)

Are you making use of [lilv](http://drobilla.net/docs/lilv/) for the LV2 support?

**Comment by grejppi** [comment]
@simonvanderveldt There's no other way than Lilv really

**Comment by unfa** [comment]
Is this feature seeing any progress? How can I help it?

**Comment by grejppi** [comment]
Can't do much until 1.2 is out, sorry

**Comment by Spekular** [comment]
@grejppi Now that there's a separate 1.2-stable branch, could work proceed for this in master? Or is the holdup something else related to 1.2?

**Comment by falkTX** [comment]
I don't want to sound too negative, but redoing the plugin system just for lv2 support seems a bit overkill.
If there's plans to add other plugin formats, then I guess it might be worth it, but just for LV2...

Is LADSPA and WinVST still working on your code, @badosu ?

**Comment by grejppi** [comment]
@Spekular Yes, it might be a good time.

In my plan LV2 support is tied with #2616, which eventually would lead to a new implementation of LADSPA effects as well. If a new (potentially incompatible) effect system is going to be in place for a future release, that might be a good opportunity to consider the proposed 0..1 range for wet/dry knobs as well. (#3422)

@falkTX It's not just for LV2, it's also for the others. LADSPA and LV2 are very similar on API level, but VST is also close on a conceptual level. If VST3 or something else is to be considered, implementing that would be easier because there won't be so many wheels to be reinvented.

**Comment by badosu** [comment]
@falkTX I was never able to progress on this, at the time I was more ambitious than I was able to deliver. My illustration above was for a side-project to understand how lv2 hosts work, it was not even embedded inside LMMS.

**Comment by faiyadesu** [comment]
> @simonvanderveldt There's no other way than Lilv really

im working on this using lilv (the c++ header)
I have a PR in my fork, will open a PR here when i have a demo :+1:

**Comment by JohannesLorenz** [comment]
We now have an [Lv2 project](https://github.com/LMMS/lmms/projects/10) and a [clean PR for a core implementation](https://github.com/LMMS/lmms/pull/4899).

The problem is not the lack of coders, it's the lack of **code reviewers**. Any volunteers?
Note that a [code review](https://github.com/LMMS/lmms/wiki/Reviewing-Pull-Requests) can always be split up, e.g. functional vs style.

**Comment by musikBear** [comment]
@JohannesLorenz IF the style-conventions are cleanly and thoroughly stated, and exists in a document, i will make **style**-review.

**Comment by Spekular** [comment]
@musikBear the "code review" link in Johannes latest comment goes to this page: https://github.com/LMMS/lmms/wiki/Reviewing-Pull-Requests

Which describes how code review is done. From there, there's a link to the coding conventions:
https://github.com/LMMS/lmms/wiki/Coding-conventions

Which state the style conventions, cleanly and thoroughly, in a document

**Comment by musikBear** [comment]
@Spekular Thanks!
Sadly it look like i cant see any code *online*.
I get this exception when i try to open the page 
**Please note that GitHub no longer supports your web browser.**
I cant use *git* either.
Looks like i cant assist before i get new hardware
Sorry!

**Comment by JohannesLorenz** [comment]
Lv2 core is in master now (2a66e83).

Some stats about the Lv2 core PR, just for fun:

* 434 days PR open
* 299 comments
* 155 commits
* 39 files changed
* 6 reviewers

Thanks to everyone making this possible!

**Comment by JohannesLorenz** [comment]
Meanwhile, the Lv2 URID feature PR is ready for review: #5517 .

**Comment by badosu** [comment]
Great job all involved!

**Comment by JohannesLorenz** [comment]
Lv2 URID feature is now supported on master (7a9b336). Next will probably be more atom support (events for piano-based instruments) or the options extension.

**Comment by JohannesLorenz** [comment]
Lv2 has MIDI support (using Atoms) now (8b2902c27a03d5678b9571ae10977b75987d44ce). This enables instruments (like pianos or guitars).

Next things will be options and buf-size. Then maybe worker.

**Comment by JohannesLorenz** [comment]
Major news of the last weeks (thanks to the reviewers, especially @DomClark and @zonkmachine):

* Lv2 Options are implemented
* Logarithmic ports have logarithmic (well, fake-logarithmic - #5344) scales now

Possible next things (though this should wait until reorg #5592 is finished):

* Support for more Parameters
* Lv2 Worker threads
* Enable CV ports (by using interpolation)

**Comment by zonkmachine** [comment]
Also Buf Size is partly done.

**Comment by djnotes** [comment]
Is LV2 usable in LMMS now? I have configured LMMS build and it shows:  
```
* Lv2 plugins                 : not found, install it or set PKG_CONFIG_PATH appropriately
```
Should I install a specific package (on Fedora) to make this warning go away?

**Comment by JohannesLorenz** [comment]
Do you have the packages "lv2" and "lilv" installed (on some distributions, you may need source RPMs for them, too)? You need both in order to compile LMMS with Lv2 support.

**Comment by Gps2010** [comment]
Interesting, I need to do some testing on openSUSE.
I do not have Lv2 nor Lilv installed.

LMMS did compile without errors, but I think that's because I have, liblilv installed. 

a C library to make use of LV2 plugins. 

In yast , when looking for lv2, I also found a long list with things like lv2 calf plugins.

**Comment by djnotes** [comment]
In fact I think lv2 or lilv are not dependencies for LMMS and currently the show as optional in the cmake log. However, I installed lv2* and lilv* and that "not found" was changed to OK.

**Comment by Gps2010** [comment]
I have them installed now, so will have to compile again.

I am also trying out Pipe Wire, and currently have an issue. When I checked I noticed jack not running though, and LMMS did not produce any sound. Setting LMMS to jack did not fix it, and every time I start LMMS it now shows me the settings screen.
So LMMS is not happy at the moment.

I might know why though, I probably need to install pipe wire jack something.

When I installed Calf lv2 plugins and those two files, YAST wanted to install jack, which I accepted.
I hope pipe wire will live up to its promises, but it does not seem to be there yet.
With pipe wire pulse, and LMMS set to pulse, LMMS did produce sound. ( even though pulse is not installed anymore at my pc)

**Comment by JohannesLorenz** [comment]
Lv2 support should be completely independent of the backend used. If Yast forces you to install jack, that's really strange.

Please note, this issue is mostly for users requesting specific Lv2 features or for devs updating about new Lv2 development. I think build issues, even if related to Lv2, should go into a separate issue (or even better, discord chat).

**Comment by Gps2010** [comment]
I have some thinking to do. Installing pipe wire removes pulse which makes sense, but I fear this will mess up my compiling the LMMS master branch. I am gonna bother the openSUSE forum about this.

**Comment by Gps2010** [comment]
Can I ask a question, because I am getting confused ?

Can somebody tell me if the newer Calf plugins should work in LMMS, and should show their Gui yet? 

I think I can load them but the specific calf gui is not showing.

For those wondering what gui I am talking about:
https://www.youtube.com/watch?v=u_PR46u1T_s&list=PLeRYlL4iVEr7R8hQnlz2G2LOacg972AeI&t=783s

**Comment by zonkmachine** [comment]
> Can somebody tell me if the newer Calf plugins should work in LMMS, and should show their Gui yet?

No gui support yet via lmms native lv2. This works with Carla though. Some plugins are blacklisted for now for various reasons, most commonly because they simply wont work without the gui but also other issues. Some Calf plugins are among those blacklisted. They can be enabled though. See this comment on startup:
```
Lv2 Plugins blacklisted: 67 of 744 
  If you want to ignore the blacklist (dangerous!), please set
  environment variable "LMMS_IGNORE_BLACKLIST" to nonempty.
```

**Comment by Gps2010** [comment]
Thank you :)

**Comment by mxmilkiib** [comment]
https://github.com/calf-studio-gear/calf/issues/248

**Comment by JohannesLorenz** [comment]
Some updates of the recent months of Lv2 development:

* We fixed #6401
* #6484 is in follow-up review (still waiting there for 6 months now)
* Development on #6707 has started

**Comment by zonkmachine** [comment]
> * is in follow-up review (still waiting there for 6 months now)

#6484 is right now a discussion about buffer sizes between two of the more experienced developers. Is that a necessary discussion to help finishing the PR?

**Comment by mxmilkiib** [comment]
FWIW, I am reminded there is the host compatability list - https://lv2plug.in/pages/host-compatibility.html - https://gitlab.com/lv2/site/-/blob/main/host_compatibility.ttl
